fish-rust/src/curses.rs

```diff
@@ -0,0 +1,261 @@
+mod sys {
+    pub const OK: i32 = 0;
+    pub const ERR: i32 = -1;
+
+    extern "C" {
+        /// The ncurses `cur_term` TERMINAL pointer.
+        pub static mut cur_term: *const core::ffi::c_void;
+
+        /// setupterm(3) is a low-level call to begin doing any sort of `term.h`/`curses.h` work.
+        /// It's called internally by ncurses's `initscr()` and `newterm()`, but the C++ code called
+        /// it directly from [`initialize_curses_using_fallbacks()`].
+        pub fn setupterm(
+            term: *const libc::c_char,
+            filedes: libc::c_int,
+            errret: *mut libc::c_int,
+        ) -> libc::c_int;
+
+        /// Frees the `cur_term` TERMINAL  pointer.
+        pub fn del_curterm(term: *const core::ffi::c_void);
+
+        /// Checks for the presence of a termcap flag identified by the first two characters of
+        /// `id`. The C function returns an integer, but we just reinterpret that as a bool.
+        pub fn tgetflag(id: *const libc::c_char) -> bool;
+
+        /// Checks for the presence and value of a number capability in the termcap/termconf
+        /// database. A return value of `-1` indicates not found.
+        pub fn tgetnum(id: *const libc::c_char) -> libc::c_int;
+
+        pub fn tgetstr(
+            id: *const libc::c_char,
+            area: *mut *mut libc::c_char,
+        ) -> *const libc::c_char;
+    }
+}
+
+// String capabilities
+pub const ENTER_ITALICS_MODE: StringCap = StringCap::new("ZH");
+pub const EXIT_ITALICS_MODE: StringCap = StringCap::new("ZR");
+pub const ENTER_DIM_MODE: StringCap = StringCap::new("mh");
+
+// Number capabilities
+pub const MAX_COLORS: Number = Number::new("Co");
+
+// Flag capabilities
+pub const EAT_NEWLINE_GLITCH: Flag = Flag::new("xn");
+
+pub struct Term {
+    has_overrides: AtomicBool,
+    overrides: Mutex<Vec<(StringCap, Arc<CString>)>>,
+}
+
+impl Term {
+    /// Internal constructor function. Like `Default` but only usable from within the module.
+    fn new() -> Self {
+        Term {
+            has_overrides: AtomicBool::new(false),
+            overrides: Mutex::new(Vec::new()),
+        }
+    }
+
+    /// Looks up support for [`Capability`] `capability` in the termcap/terminfo database via the
+    /// curses library.
+    pub fn get<'a, C: Capability<'a>>(&'a self, capability: C) -> C::Result {
+        capability.lookup(self)
+    }
+
+    /// Overrides the string value of `capability` for the current terminal.
+    pub fn set(&self, id: StringCap, value: &str) {
+        let value = CString::new(value).unwrap().into();
+        let mut overrides = self.overrides.lock().expect("Mutex poisoned!");
+        match overrides.binary_search_by(|entry| entry.0.cmp(&id)) {
+            Ok(idx) => overrides[idx] = (id, value),
+            Err(idx) => overrides.insert(idx, (id, value)),
+        }
+        self.has_overrides.store(true, Ordering::Relaxed);
+    }
+}
+
+pub trait Capability<'a> {
```

Comment:
Adding a function for each feature instead of passing the name of the capability to look up is just less rusty. And I don't want to add a getter *and* a setter for each capability we might want to override. Making the capability just a parameter to get/set is a lot less tedious and the ergonomics don't suffer for it.